vmm/src/vm.rs

```diff
@@ -285,6 +288,16 @@ impl Vcpu {
             },
         }
     }
+
+    fn patch_cpuid(&self, cpuid: &mut CpuId) {
```

Comment:
Could you work out how to refactor this to reuse the existing `patch_cpuid` function in vm.rs?

Reply:
It's actually very hard to reuse existing patch_cpuid, which only takes care of setting one bit while here we need update a whole EDX. 
Actually I think the abstract of CpuidPatch is not flexible and over-killed here. Considering we will only have very limited number of CPUID patching, a simple 'match' block would already be clear and have good performance.

Note the patch_cpuid() I'm adding is per-vCPU level while the existing one is per-VM level. This is necessary since APIC ID is specific to a certain vCPU. I tend to keep the per-vCPU one(e.g. this patch) and keep or revert the per-VM one (that @sboeuf  added).

Let me know your ideas.

Reply:
Maybe don't use such a generic name for a such a special purpose function

Reply:
> Actually I think the abstract of CpuidPatch is not flexible and over-killed here. 

I feel that the existing function:

```Rust
fn patch_cpuid(cpuid: &mut CpuId, patches: Vec<CpuidPatch>) {
```

is actually very generic and could be extracted out of the `Vm` implementation to be used for both the per vCPU and the VM CPUID patching.

@chao-p Provided that `patch_cpuid` would be a standalone routine, what would prevent you from using it with a single entry `patches` array?

Reply:
@sameo The problem with  patch_cpuid/ CpuidPatch is it can only set one bit at a time if you look at its implementation and that is really the simplest case and even does not satisfy this example. If we really want to make it generic, we need at least support below functionalities:
- set certain bit(s)
- clear certain bit(s)
- override certain bit(s)

However I see the code will become complex once we do that and even we can do that we will still have performance drawback than simple 'match' block. When the number of CPUID being patched is small, I'd choose the simple and powerful 'match' block: all the above functionalities can be done with very simple code.

Reply:
> When the number of CPUID being patched is small, I'd choose the simple and powerful 'match' block

What do you mean by "match block"

> all the above functionalities can be done with very simple code.

I'm asking for common code here. Sebastien's code and what you're proposing have almost identical goals. If there's a simple implementation for both then we should have one.

Reply:
I mean using one loop and in the loop then using match to patch all the CPUID one by one. That's basically the existing code in crosvm and firecracker if you look at their filter_cpuid().

Reply:
> I mean using one loop and in the loop then using match to patch all the CPUID one by one. That's basically the existing code in crosvm and firecracker if you look at their filter_cpuid().

AFAIU `filter_cpuid()`  statically modifies/generates CPUID, it does not patch it dynamically like what we're trying to do here.

Reply:
Not really. They did the same thing that we need to do and these info need to be updated dynamically anyway since you can't know a vCPU's APIC ID statically.

Reply:
```Rust
pub fn filter_cpuid(
    cpu_id: u8,
    cpu_count: u8,
    ht_enabled: bool,
    kvm_cpuid: &mut CpuId,
```

Yes, `filter_cpuid` does what you did and set the cpu ID properly. But it does not allow us to change bit by bit like what Sebastien needed.
I'd suggest to rename your routine to incorporate the fact that it only sets the cpu ID.

Reply:
Yes at this time the function only sets the APIC ID but it's expected to to have other vCPU-level CPUID patches to be added to this function as well. But sure if you think this is the right way I can change the name to update_apicid.

Reply:
Ok so I completely understand the patching of the APIC ID has to happen right before the CPUID is set, and even more than that, it has to happen from a VCPU granularity compared to the generic CPUID that was generated at the VM level.
That's why I think it's actually okay to have both `patch_cpuid()` implemented by `Vcpu` and by `Vm`. One comment I have would be to rename it to something like patch_cpuid_reg() because you're patching an entire register at once.
Also you could make this function a bit more generic, so that it could work for any register.

Reply:
> Also you could make this function a bit more generic, so that it could work for any register.

@chao-p Yes, something like `set_cpuid_reg(cpuid, function, reg, value)`

Reply:
Updated the PR with the above suggestion, please take a look.